### PR TITLE
Revert "miri: make sure we can find link_section statics even for the local crate"

### LINKED ISSUE #127052: Regression in `#[used]` attribute on Windows msvc
### Code

I tried this code:

```rust
#[used]
static FOO: u32 = 0;
fn main() {}
```

I expected to see this happen: Compiles successfully.

Instead, this happened: Fails with this error:

```
  = note: symbols.o : error LNK2001: unresolved external symbol _ZN3foo3FOO17h3b276d94f7182e74E
          D:\Temp\foo\target\debug\deps\foo.exe : fatal error LNK1120: 1 unresolved externals
```

### Version it worked on

nightly-2024-06-26 and earlier.

### Version with regression

nightly-2024-06-27 and later, starting with #126938.
See https://github.com/rust-lang/rust/pull/126938#issuecomment-2192434302 for a discussion of the issue.

`rustc --version --verbose`:
```
rustc 1.81.0-nightly (4bc39f028 2024-06-26)
binary: rustc
commit-hash: 4bc39f028d14c24b04dd17dc425432c6ec354536
commit-date: 2024-06-26
host: x86_64-pc-windows-msvc
release: 1.81.0-nightly
LLVM version: 18.1.7
```

cc @RalfJung @ChrisDenton

### PR DESCRIPTION
**Pull request by lqd**
This PR reverts #126938 as [requested by its author](https://github.com/rust-lang/rust/issues/127052#issuecomment-2196793473), to fix the #127052 regression.

Fixes #127052

We should probably improve the [`used` rmake test(s)](https://github.com/rust-lang/rust/blob/57931e50409f9365791f7923a68f9ae1ec301c4c/tests/run-make/used/rmake.rs#L7) in the future, but this should do for now.

### CODE REVIEW

**Comment by workingjubilee** [comment]
Thank you for filing this revert + regression test!

**Review by compiler-errors — APPROVED** [review]
thanks @lqd, r=me when this is done bors-trying so we can fix this

**Comment by lqd** [comment]
The test [looks good](https://github.com/rust-lang-ci/rust/actions/runs/9718507842/job/26826546504#step:25:18286) on the `try` build already.

> test [ui] tests\ui\linkage-attr\unreferenced-used-static-issue-127052.rs ... ok

(and by now an old windows box I have has finished rebuilding LLVM to workaround the recent linking issues people have experienced, so I also checked that the test does fail on master without the revert -- no need for another try build 🚀).

I'm not sure if `//@ only-msvc` is restrictive enough in this case (but probably?), and I guess the test comment could be more explicit about what happens but nbd -- so I wanted to ask @ChrisDenton but it seems fine for a revert :)

And if this is indeed what [caused CI to fail](https://github.com/rust-lang/rust/issues/127052#issuecomment-2197674448) recently maybe we can bump the prio.

**Comment by lqd** [comment]
finallyyyyyyy 😴 

@bors r=compiler-errors

File: tests/ui/linkage-attr/unreferenced-used-static-issue-127052.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// This is a non-regression test for issue #127052 where unreferenced `#[used]` statics couldn't be
+// removed by the MSVC linker, causing linking errors.
+
+//@ build-pass: needs linking
+//@ only-msvc
```
Comment (RalfJung) [inline tests/ui/linkage-attr/unreferenced-used-static-issue-127052.rs:5]: Why shouldn't we run this test on all targets?

Comment (lqd) [reply tests/ui/linkage-attr/unreferenced-used-static-issue-127052.rs:5]: The specific issue doesn't exist elsewhere so it's less useful there: the existing tests already cover `#[used]` features and targets more than this test. Of course, it wouldn't hurt. Feel free to change the directive to something you prefer.

Comment (RalfJung) [reply tests/ui/linkage-attr/unreferenced-used-static-issue-127052.rs:5]: Well, clearly the existing tests do not cover `#[used]` in the binary, or my PR would not have landed.

**Comment by apiraino** [comment]
thank you @lqd :bowing_man: 

(sometimes I have to spell it out where a emoji is not enough)

**Pull request merged**